apps/user-profile/src/app/pages/settings-page/settings-authorization/settings-authentication.component.ts

```diff
@@ -217,9 +206,314 @@ export class SettingsAuthenticationComponent implements OnInit, AfterViewInit {
+          this.showDetail = !this.showDetail;
+        })
+      );
+    } else {
+      this.showDetail = !this.showDetail;
+    }
+  }
+
+  toggleEnableMfa(): void {
+    this.unchangedEnforce = false;
+    this.includeCategories = [];
+    this.allCategories = !this.toggle.checked;
+    this.setValuesFromSetting();
+  }
+
+  toggleCategory(categoryValue: any): void {
+    this.unchangedSettings = false;
+    categoryValue.value = !categoryValue.value;
+    for (const rps in categoryValue.rps_value) {
+      categoryValue.rps_value[rps] = categoryValue.value;
+    }
+  }
+
+  toggleRps(rpsValue: string): boolean {
+    this.unchangedSettings = false;
+    return !rpsValue;
+  }
+
+  /**
+   * This method fires logic for setting new values of enforceMfa and settings
+   */
+  saveSettings(enforceFirstMfa = false): void {
+    if (this.oauthService.getIdTokenExpiration() - now() > 0 && !enforceFirstMfa) {
+      this.loadingMfa = true;
+      if (this.enforceMfa !== this.toggle.checked) {
+        this.enableMfa(this.toggle.checked);
+      }
+      if (!this.unchangedSettings) {
+        this.saveDetailSettings();
+        this.updateDetailSettings();
+      }
+    } else {
+      this.saveEnforceMfa();
+      if (!this.unchangedSettings) {
+        this.saveDetailSettings();
+      }
+      this.reAuthenticate();
+    }
+  }
+
+  saveEnforceMfa(): void {
+    if (this.enforceMfa !== this.toggle.checked) {
+      sessionStorage.setItem('enforce_mfa', this.toggle.checked.toString());
+    }
+  }
+
+  /**
+   * This method creates request body for new settings according to toggles
+   */
+  saveDetailSettings(): void {
+    let allTrueCheck = true;
+    let allFalseCheck = false;
+    this.includeCategories = [];
+    this.excludeRps = [];
+
+    for (const category in this.categories) {
+      if (this.categories[category].value) {
+        allFalseCheck = true;
+        this.includeCategories.push(category);
+      } else {
+        allTrueCheck = false;
+        continue;
+      }
+      for (const rps in this.categories[category]['rps']) {
+        if (this.categories[category].rps_value[rps]) {
+          allFalseCheck = true;
+        } else {
+          allTrueCheck = false;
+          this.excludeRps.push(rps);
+        }
+      }
+    }
+
+    let body: string;
+
+    if (allTrueCheck === allFalseCheck) {
+      if (allTrueCheck) {
+        body = JSON.stringify({ all: true });
+      } else {
+        body = '{}';
+      }
+    } else {
+      body = JSON.stringify({
+        include_categories: this.includeCategories,
+        exclude_rps: this.excludeRps,
+      });
+    }
+
+    sessionStorage.setItem('settings_mfa', body);
+  }
+
+  /**
+   * This method is used when we want to do some PUT operation, but we need new (not expired) id token
+   */
+  reAuthenticate(): void {
+    sessionStorage.setItem('mfa_route', '/profile/settings/auth');
+    localStorage.removeItem('refresh_token');
+    this.oauthService.logOut(true);
```

Comment:
since we are storing all the tokens in the localStorage the method logOut already remove refresh token
```suggestion
    this.oauthService.logOut(true);
```